internal/controller/stages/regular_stages.go

```diff
@@ -0,0 +1,1981 @@
+func (r *RegularStagesReconciler) syncPromotions(
+	ctx context.Context,
+	stage *kargoapi.Stage,
+) (kargoapi.StageStatus, bool, error) {
+	logger := logging.LoggerFromContext(ctx)
+	newStatus := *stage.Status.DeepCopy()
+
+	// List all Promotions for the Stage.
+	promotions := &kargoapi.PromotionList{}
+	if err := r.client.List(
+		ctx,
+		promotions,
+		client.InNamespace(stage.Namespace),
+		client.MatchingFieldsSelector{
+			Selector: fields.OneTermEqualSelector(indexer.PromotionsByStageField, stage.Name),
+		},
+	); err != nil {
+		err = fmt.Errorf(
+			"failed to list Promotions for Stage %q in namespace %q: %w",
+			stage.Name, stage.Namespace, err,
+		)
+
+		conditions.Set(&newStatus, &metav1.Condition{
+			Type:               kargoapi.ConditionTypePromoting,
+			Status:             metav1.ConditionUnknown,
+			Reason:             "ListPromotionsFailed",
+			Message:            err.Error(),
+			ObservedGeneration: stage.Generation,
+		})
+
+		return newStatus, false, err
+	}
+
+	// If there are no Promotions, then we are not promoting any Freight.
+	if len(promotions.Items) == 0 {
+		logger.Debug("no Promotions found for Stage")
+
+		// Ensure we delete any existing "current" Promotion related information
+		// from the Stage status.
+		conditions.Delete(&newStatus, kargoapi.ConditionTypePromoting)
+		newStatus.CurrentPromotion = nil
+
+		return newStatus, false, nil
+	}
+
+	// Sort the Promotions by phase and creation time to determine the current
+	// state of the Stage.
+	slices.SortFunc(promotions.Items, kargoapi.ComparePromotionByPhaseAndCreationTime)
+
+	// The Promotion with the highest priority (i.e. a Running or Pending phase)
+	// is the one that we will consider for the current state of the Stage.
+	highestPrioPromo := promotions.Items[0]
+
+	// The Promotion which is currently running on the Stage.
+	currentPromo := stage.Status.CurrentPromotion
+
+	// The last Promotion which ran on the Stage.
+	lastPromo := stage.Status.LastPromotion
+
+	// Track if there are any non-terminal promotions that need handling.
+	// This is later used to determine if we should issue an immediate
+	// requeue.
+	var hasNonTerminalPromotions bool
+	for _, promo := range promotions.Items {
+		if !promo.Status.Phase.IsTerminal() {
+			hasNonTerminalPromotions = true
+			break
+		}
+	}
+
+	// If the current Promotion is not the highest priority Promotion, or the
+	// highest priority Promotion is in a terminal phase, then we must have
+	// finished promoting.
+	if currentPromo != nil && (currentPromo.Name != highestPrioPromo.Name || highestPrioPromo.Status.Phase.IsTerminal()) {
+		// Gather information about the Promotions that have terminated since the
+		// last reconciliation.
+		var newPromotions []kargoapi.PromotionReference
+		for _, promo := range promotions.Items {
+			// Update the conditions to reflect that we are no longer promoting.
+			conditions.Delete(&newStatus, kargoapi.ConditionTypePromoting)
+			newStatus.CurrentPromotion = nil
```

Comment:
Should this be done just once before the loop starts?

Reply:
If we delete it at the beginning by default, the transition timestamp on the condition will continue to drift, which is why they are removed for individual cases.